snuba/query/processors/failure_rate_processor.py

```diff
@@ -33,7 +33,33 @@ def process_functions(exp: Expression) -> Expression:
                         binary_condition(
                             None,
                             ConditionFunctions.NOT_IN,
-                            Column(None, None, "transaction_status"),
+                            # Clickhouse's NOT IN operator errors when the left hand side
+                            # is explicity NULL while the right hand side contains UInt8.
+                            # However it works when there the left hand side is a NULL
+                            # generated by the result of some operation (without explicitly
+                            # using NULL)
+                            #
+                            # e.g. SELECT NULL in (1,2)                 <- gives error
+                            #      SELECT toUInt8OrNull('') in (1,2)    <- gives NULL
+                            #
+                            # So if transaction_status is not NULL, we can directly use it,
+                            # but when it is NULL, we coalesce a NULL value using toUInt8OrNull.
+                            #
+                            # This permits the failure_rate function to not error when making
+                            # queries to tables without transaction_status (e.g. events table)
```

Comment:
This seems a very weak approach. I think Clickhouse happens to behave this way by coincidence and this behavior could be fixed (which would make this stop working).
if you need the entire function to return NULL when `transaction_status` is NULL, why not actually make this condition explicit?
https://clickhouse.tech/docs/en/sql-reference/functions/conditional-functions/#if
```
if(transaction_status IS NULL, NULL, divide(......))
```

This way, while more verbose, it would be a lot more robust and readable:
1) we would not be relying on an unspecified Clckhouse behavior
2) we would not have to make the mental leap that by returning NULL in the NOT IN condition we are making the external function return NULL.
3) You would not have to justify in comments why this works, the code would be self explanatory.

Reply:
Unfortunately Clickhouse's conditional function does not short circuit. From the [docs](https://clickhouse.tech/docs/en/sql-reference/functions/conditional-functions/#if)

> Unlike most systems, ClickHouse always evaluate both expressions then and else.

This behavior means that if `transaction_status` is NULL, the query will still error when evaluating the else expression even though we only want the results of the then expression. So we're still facing the same issue.

Clickhouse's behavior with NULL seems to usually result in another NULL, as such I'd like to believe that `SELECT NULL IN (1,2)` causing an error is a bug (though I do agree this is weak). So if we'd like to not rely on the `NULL` behavior of the `IN` operator, another approach I can think of is to leverage `OR`s and `equals` like

```
if(or(equals(transaction_status, 0), or(equals(transaction_status, 1), equals(transaction_status, 2))))
```

In this case the comparison function will return NULL when `transaction_status` is NULL. This behavior is documented behavior here in the [docs](https://clickhouse.tech/docs/en/sql-reference/syntax/#null-literal)

> if at least one of the arguments of a comparison operation is `NULL`, the result of this operation is also `NULL`

I don't think we can do something similar with the `IN` operator as we can observe that the NULL behavior is inconsistent.

Reply:
I think the OR approach would be ok. Well, likely the best we ca do. At least this is clearly documented.